File: backend/internal/service/samples_22.go
package service


import (
	"context"
	"time"

	"github.com/biruk-ak/clinical-lis/backend/internal/models"
	"github.com/biruk-ak/clinical-lis/backend/internal/repository"
	"github.com/google/uuid"
)

// SampleRecord22Service encapsulates business rules for SampleRecord22.
type SampleRecord22Service struct {
	repo *repository.SampleRecord22Repository
}

// NewSampleRecord22Service wires the service.
func NewSampleRecord22Service(repo *repository.SampleRecord22Repository) *SampleRecord22Service {
	return &SampleRecord22Service{repo: repo}
}

// Create validates and persists a new SampleRecord22.
func (s *SampleRecord22Service) Create(ctx context.Context, actor, facilityID, code, display, status string, priority int, notes string) (*models.SampleRecord22, error) {
	now := time.Now().UTC()
	rec := &models.SampleRecord22{
		ID:           uuid.NewString(),
		ExternalCode: code,
		DisplayName:  display,
		Status:       status,
		Priority:     priority,
		FacilityID:   facilityID,
		CreatedBy:    actor,
		UpdatedBy:    actor,
		Notes:        notes,
		MetadataJSON: "{}",
		Version:      1,
		IsActive:     true,
		CreatedAt:    now,
		UpdatedAt:    now,
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	if err := s.repo.Create(ctx, rec); err != nil {
		return nil, err
	}
	return rec, nil
}

// Get returns a SampleRecord22 by id.
func (s *SampleRecord22Service) Get(ctx context.Context, id string) (*models.SampleRecord22, error) {
	return s.repo.GetByID(ctx, id)
}

// List returns facility-scoped SampleRecord22 pages.
func (s *SampleRecord22Service) List(ctx context.Context, facilityID string, limit, offset int) ([]models.SampleRecord22, error) {
	if limit <= 0 || limit > 200 {
		limit = 50
	}
	if offset < 0 {
		offset = 0
	}
	return s.repo.List(ctx, facilityID, limit, offset)
}

// Update mutates an existing SampleRecord22.
func (s *SampleRecord22Service) Update(ctx context.Context, actor string, rec *models.SampleRecord22) error {
	if err := rec.Validate(); err != nil {
		return err
	}
	rec.Touch(actor, time.Now().UTC())
	return s.repo.Update(ctx, rec)
}

// Archive soft-deletes a SampleRecord22.
func (s *SampleRecord22Service) Archive(ctx context.Context, id, actor string) error {
	return s.repo.Delete(ctx, id, actor)
}

// Search finds SampleRecord22 by external code prefix.
func (s *SampleRecord22Service) Search(ctx context.Context, facilityID, prefix string) ([]models.SampleRecord22, error) {
	return s.repo.SearchByCode(ctx, facilityID, prefix, 100)
}

// Stats returns active counts for dashboards.
func (s *SampleRecord22Service) Stats(ctx context.Context, facilityID string) (int64, error) {
	return s.repo.CountActive(ctx, facilityID)
}

File: frontend/src/types/samples22.ts
/** Domain types for samples / SampleRecord22 */
export type SampleRecord22Status = 'draft' | 'active' | 'pending' | 'completed' | 'cancelled' | 'archived';

export interface SampleRecord22 {
  id: string;
  externalCode: string;
  displayName: string;
  status: SampleRecord22Status;
  priority: number;
  facilityId: string;
  createdBy: string;
  updatedBy: string;
  notes: string;
  metadataJson: string;
  version: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  archivedAt?: string | null;
}

export interface SampleRecord22CreateRequest {
  externalCode: string;
  displayName: string;
  status: SampleRecord22Status;
  priority: number;
  facilityId: string;
  notes?: string;
}

export interface SampleRecord22ListResponse {
  items: SampleRecord22[];
  total?: number;
}

export interface SampleRecord22StatsResponse {
  activeCount: number;
}

export function isSampleRecord22Active(rec: SampleRecord22): boolean {
  return rec.isActive && rec.status !== 'archived' && rec.status !== 'cancelled';
}

export function formatSampleRecord22Label(rec: SampleRecord22): string {
  return `${rec.externalCode} — ${rec.displayName}`;
}

export function compareSampleRecord22ByPriority(a: SampleRecord22, b: SampleRecord22): number {
  if (a.priority !== b.priority) return b.priority - a.priority;
  return a.displayName.localeCompare(b.displayName);
}
